File: python/packet.py
import struct
from enum import IntEnum
from typing import List, Optional

from crc import Calculator, Crc8


class Entry:

    def __init__(self, name: str):
        self.name = name
        self.type_ = 0b000000
        self.size = 0
        self.payload = bytes()
        self.sub_entries = []
        self.sub_packet = None

    def is_null(self) -> bool:
        return self.match_type(0b000000)

    def is_int(self) -> bool:
        return self.match_type(0b010000, 0b110000) or self.match_type(
            0b100000, 0b100000
        )

    def is_float(self) -> bool:
        return self.match_type(0b000100, 0b111100)

    def is_float16(self) -> bool:
        return self.match_type(0b000101)

    def is_float32(self) -> bool:
        return self.match_type(0b000110)

    def is_float64(self) -> bool:
        return self.match_type(0b000111)

    def is_bytes(self) -> bool:
        return self.match_type(0b000011) or self.match_type(0b001000, 0b111000)

    def is_struct(self) -> bool:
        return self.match_type(0b000001)

    def is_packet(self) -> bool:
        return self.match_type(0b000010)

    def bool(self) -> bool:
        return bool(self.int())

    def int(self) -> int:
        if self.match_type(0b010000, 0b110000):
            is_negative = self.type_ & 0b001000
            buf = bytearray(8)
            buf[0 : self.size] = self.payload
            value = struct.unpack("<Q", buf)[0]
            if is_negative:
                return -value
            else:
                return value
        if self.match_type(0b100000, 0b100000):
            return self.type_ & 0b011111

        return 0

    def float(self) -> float:
        if self.match_type(0b000100, 0b111111):
            return 0.0
        if self.is_float16():
            return struct.unpack("<e", self.payload)[0]
        if self.is_float32():
            return struct.unpack("<f", self.payload)[0]
        if self.is_float64():
            return struct.unpack("<d", self.payload)[0]
        if self.is_int():
            return float(self.int())

        return 0.0

    def bytes(self) -> bytes:
        if self.match_type(0b000011):
            return self.payload[1:]
        if self.match_type(0b001000, 0b111000):
            return self.payload
        return bytes()

    def string(self) -> str:
        return self.bytes().decode()

    def packet(self) -> Optional["Packet"]:
        return self.sub_packet

    def struct(self) -> List["Entry"]:
        return self.sub_entries

    def set_null(self) -> "Entry":
        self.type_ = 0b000000
        self.size = 0
        self.payload = bytes()
        return self

    def set_bool(self, value: bool) -> "Entry":
        self.set_int(int(value))
        return self

    def set_int(self, value: int) -> "Entry":
        value = int(value)
        if value >= 0 and value < 32:
            self.type_ = 0b100000 | value
            self.size = 0
            self.payload = bytes()
        else:
            self.payload = struct.pack("<Q", abs(value))
            self.size = 8
            while self.size > 0:
                if self.payload[self.size - 1] == 0:
                    self.size -= 1
                else:
                    break
            self.payload = self.payload[: self.size]
            if value > 0:
                self.type_ = 0b010000 | (self.size - 1)
            else:
                self.type_ = 0b011000 | (self.size - 1)

        return self

    def set_float16(self, value: float) -> "Entry":
        if value == 0.0:
            self.type_ = 0b000100
            self.size = 0
            self.payload = bytes()
        else:
            self.type_ = 0b000101
            self.size = 2
            self.payload = struct.pack("<e", value)
        return self

    def set_float32(self, value: float) -> "Entry":
        if value == 0.0:
            self.type_ = 0b000100
            self.size = 0
            self.payload = bytes()
        else:
            self.type_ = 0b000110
            self.size = 4
            self.payload = struct.pack("<f", value)
        return self

    def set_float64(self, value: float) -> "Entry":
        if value == 0.0:
            self.type_ = 0b000100
            self.size = 0
            self.payload = bytes()
        else:
            self.type_ = 0b000111
            self.size = 8
            self.payload = struct.pack("<d", value)
        return self

    def set_bytes(self, data: bytes) -> "Entry":
        if len(data) <= 7:
            self.type_ = 0b001000 | len(data)
            self.size = len(data)
            self.payload = data[:]
        else:
            self.type_ = 0b000011
            self.size = 1 + len(data)
            self.payload = bytes([len(data)]) + data[:]
        return self

    def set_string(self, data: str) -> "Entry":
        return self.set_bytes(data.encode())

    def set_struct(self, sub_entries: List["Entry"]) -> "Entry":
        self.type_ = 0b000001
        self.size = 1
        self.sub_entries = sub_entries
        return self

    def set_packet(self, sub_packet: "Packet") -> "Entry":
        self.type_ = 0b000010
        self.size = sub_packet.size
        self.sub_packet = sub_packet
        return self

    def decode(self, buf: bytes) -> int:
        self.type_ = (buf[0] >> 5) | ((buf[1] & 0b11100000) >> 2)

        self.name = chr((buf[0] & 0b00011111) + 64) + chr((buf[1] & 0b00011111) + 96)

        if self.type_ == 0b000000 or self.type_ == 0b000100 or self.type_ & 0b100000:
            self.size = 0
            # null, 0.0f, short int
        elif (self.type_ & 0b110000) == 0b010000:
            self.size = (self.type_ & 0b000111) + 1
            # int
        elif self.type_ >= 0b000101 and self.type_ <= 0b000111:
            self.size = 1 << (self.type_ & 0b000011)
            # float
        elif self.type_ == 0b000011:
            self.size = 1 + buf[2]
            # bytes
        elif self.type_ == 0b000001 or self.type_ == 0b000010:
            self.size = buf[2]
            # struct, packet
        elif (self.type_ & 0b111000) == 0b001000:
            self.size = self.type_ & 0b000111
            # short bytes
        else:
            raise ValueError("Invalid entry type")
        self.payload = buf[2 : 2 + self.size]

        if self.is_struct():
            self.sub_entries = []
            i = 1
            while i < self.size:
                entry = Entry("@@")
                i += entry.decode(self.payload[i:])
                self.sub_entries.append(entry)

        if self.is_packet():
            self.sub_packet = Packet.decode(self.payload[: self.size])

        return 2 + self.size

    def encode(self) -> bytes:
        if self.is_struct():
            self.payload = bytearray([0])
            self.size = 1
            for entry in self.sub_entries:
                entry_buf = entry.encode()
                self.payload += entry_buf
                self.size += len(entry_buf)
            self.payload[0] = self.size

        if self.is_packet():
            self.payload = self.sub_packet.encode()
            self.size = self.sub_packet.size

        buf = bytearray(2 + len(self.payload))
        buf[0] = ((self.type_ & 0b000111) << 5) | (ord(self.name[0].upper()) - 64)
        buf[1] = ((self.type_ & 0b111000) << 2) | (ord(self.name[1].upper()) - 64)
        buf[2:] = self.payload
        return buf

    def match_type(self, value, mask=0b111111):
        return (self.type_ & mask) == value

    def print(self, indent: str = ""):
        payload_str = ""
        type_str = ""
        if self.is_null():
            payload_str = "null"
            type_str = "null   "
        elif self.is_int():
            payload_str = str(self.int())
            type_str = "int    "
        elif self.is_float16():
            type_str = "float16"
            payload_str = str(self.float())
        elif self.is_float32():
            type_str = "float32"
            payload_str = str(self.float())
        elif self.is_float64():
            type_str = "float64"
            payload_str = str(self.float())
        elif self.is_bytes():
            type_str = "bytes  "
            payload_str = repr(self.string())
        elif self.is_packet():
            type_str = "packet "
        elif self.is_struct():
            type_str = "struct "

        print(indent + self.name + ": " + type_str + " = " + payload_str)

        if self.is_packet():
            self.packet().print(indent + "  ")
        elif self.is_struct():
            for entry in self.struct():
                entry.print(indent + "  ")


class PacketType(IntEnum):
    COMMAND = 0
    TELEMETRY = 1

    def __str__(self) -> str:
        if self == PacketType.COMMAND:
            return "cmd"
        else:
            return "tlm"


class Packet:
    """Packet"""

    size_max = 255
    entry_type_size = 2
    unit_id_local = 0x00
    component_id_self = 0x00
    packet_type_mask = 0b10000000
    packet_id_mask = 0b01111111

    def __init__(self):
        self.size = 0
        self.type_ = PacketType.COMMAND
        self.packet_id = 0
        self.component_id = 0
        self.origin_unit_id = 0
        self.dest_unit_id = 0
        self.sequence = 0
        self.entries: List[Entry] = []

        self.buf = None

    def is_command(self) -> bool:
        return self.type_ == PacketType.COMMAND

    def is_telemetry(self) -> bool:
        return self.type_ == PacketType.TELEMETRY

    def is_local(self) -> bool:
        return self.origin_unit_id == self.unit_id_local

    def is_remote(self) -> bool:
        return self.origin_unit_id != self.unit_id_local

    def checksum(self) -> int:
        if self.buf is None:
            return Calculator(Crc8.CCITT).checksum(self.encode())
        else:
            return Calculator(Crc8.CCITT).checksum(self.buf)

    def resize(self):
        pass

    def find(self, name: str) -> Optional[Entry]:
        for entry in self.entries:
            if entry.name.lower() == name.lower():
                return entry
        return None

    def print(self, indent: str = ""):
        if self.is_local():
            print(
                f"{indent}{self.type_}[{self.size}] {hex(self.packet_id)} by {hex(self.component_id)}, local"
            )
        else:
            print(
                f"{indent}{self.type_}[{self.size}] {hex(self.packet_id)} by {hex(self.component_id)}"
                + f", remote (from {hex(self.origin_unit_id)} to {hex(self.dest_unit_id)}), #{self.sequence}"
            )

        for entry in self.entries:
            entry.print(indent + "  ")

    def encode(self) -> bytes:
        buf = bytearray(255)
        buf[1] = (int(self.type_) * self.packet_type_mask) | (
            self.packet_id & self.packet_id_mask
        )
        buf[2] = self.component_id
        buf[3] = self.origin_unit_id
        if self.is_remote():
            buf[4] = self.dest_unit_id
            buf[5:7] = struct.pack("<H", self.sequence)
            i = 7
        else:
            i = 4
        for entry in self.entries:
            entry_buf = entry.encode()
            buf[i : i + len(entry_buf)] = entry_buf
            i += len(entry_buf)

        buf[0] = i

        return bytes(buf[:i])

    @classmethod
    def command(
        cls,
        packet_id: int = 0,
        component_id: int = 0,
        origin_unit_id: int = 0,
        dest_unit_id: int = 0,
        sequence: int = 0,
    ) -> "Packet":
        packet = cls()
        packet.type_ = PacketType.COMMAND
        packet.packet_id = packet_id
        packet.component_id = component_id
        packet.origin_unit_id = origin_unit_id
        packet.dest_unit_id = dest_unit_id
        packet.sequence = sequence
        return packet

    @classmethod
    def telemetry(
        cls,
        packet_id: int = 0,
        component_id: int = 0,
        origin_unit_id: int = 0,
        dest_unit_id: int = 0,
        sequence: int = 0,
    ) -> "Packet":
        packet = cls()
        packet.type_ = PacketType.TELEMETRY
        packet.packet_id = packet_id
        packet.component_id = component_id
        packet.origin_unit_id = origin_unit_id
        packet.dest_unit_id = dest_unit_id
        packet.sequence = sequence
        return packet

    @classmethod
    def decode(cls, buf: bytes) -> Optional["Packet"]:
        if buf[0] > len(buf):
            return None

        try:
            packet = cls()

            packet.size = buf[0]
            packet.type_ = PacketType((buf[1] & cls.packet_type_mask) >> 7)
            packet.packet_id = buf[1] & cls.packet_id_mask
            packet.component_id = buf[2]
            packet.origin_unit_id = buf[3]
            packet.dest_unit_id = buf[4] if packet.is_remote() else cls.unit_id_local
            packet.sequence = (
                struct.unpack("<H", buf[5:7])[0] if packet.is_remote() else 0
            )

            packet.buf = buf[: packet.size]

            i = 7 if packet.is_remote() else 4

            while i < packet.size:
                entry = Entry("@@")
                i += entry.decode(buf[i:])
                packet.entries.append(entry)

            return packet

        except Exception as e:
            raise e
            return None

File: cpp/Packet.cpp
#include "float16.h"
#include "packet.h"

#include <cassert>
#include <cstring>

namespace wcpp {

Entry::Name &Entry::Name::operator=(const char name[2]) {
  buf_[0] = (buf_[0] & 0b11100000) | (name[0] & 0b00011111);
  buf_[1] = (buf_[1] & 0b11100000) | (name[1] & 0b00011111);
  return *this;
}

bool Entry::Name::operator==(const char name[2]) {
  return (buf_[0] & 0b00011111) == (name[0] & 0b00011111) &&
         (buf_[1] & 0b00011111) == (name[1] & 0b00011111);
}

bool Entry::Name::operator==(Name name) {
  return *this == name.buf_;
}

char Entry::Name::operator[](int i) const {
  if (i == 0) return (buf_[0] & 0b00011111) + 64; 
  if (i == 1) return (buf_[1] & 0b00011111) + 96; 
  return '\0';
} 


Entry::Name Entry::name() const {
  return Name(entries_.buf_ + ptr_); 
}

uint8_t Entry::size() const {
  uint8_t type = getType();
  if (type == 0b000000 || type == 0b000100 || type & 0b100000)
    return 0; // null, 0.0f, short int
  if ((type & 0b110000) == 0b010000)
    return (type & 0b000111) + 1; // int
  if (type >= 0b000101 && type <= 0b000111)
    return 1 << (type & 0b000011); // float
  if (type == 0b000011)
    return 1 + entries_.buf_[ptr_ + entry_type_size]; // bytes
  if (type == 0b000001 || type == 0b000010)
    return entries_.buf_[ptr_ + entry_type_size]; // struct, packet
  if ((type & 0b111000) == 0b001000)
    return type & 0b000111; // short bytes

  return 0;
}

Entry::operator bool() const {
  return ptr_ + entry_type_size < entries_.buf_size_;
}

void Entry::setType(uint8_t type) {
  entries_.buf_[ptr_ + 0] =
    (entries_.buf_[ptr_ + 0] & 0b00011111) | ((type & 0b000111) << 5);
  entries_.buf_[ptr_ + 1] =
    (entries_.buf_[ptr_ + 1] & 0b00011111) | ((type & 0b111000) << 2);
}

uint8_t Entry::getType() const {
  return (entries_.buf_[ptr_ + 0] >> 5) |
         ((entries_.buf_[ptr_ + 1] & 0b11100000) >> 2);
}

int64_t Entry::getSignedInt() const {
  if (matchType(0b010000, 0b110000)) {
    bool is_negative = getType() & 0b001000;
    if (is_negative)
      return - (int64_t)getPayload<uint64_t>((getType() & 0b000111) + 1);
    else
      return getPayload<uint64_t>((getType() & 0b000111) + 1);
  }
  if (matchType(0b100000, 0b100000))
    return getType() & 0b011111;

  return 0;
}

uint64_t Entry::getUnsignedInt() const {
  if (matchType(0b010000, 0b110000)) {
    bool is_negative = getType() & 0b001000;
    if (is_negative)
      return - (int64_t) getPayload<uint64_t>((getType() & 0b000111) + 1);
    else
      return getPayload<uint64_t>((getType() & 0b000111) + 1);
  }
  if (matchType(0b100000, 0b100000))
    return getType() & 0b011111;

  return 0;
}

float16 Entry::getFloat16() const {
  if (matchType(0b000100, 0b111111))
    return float16();
  if (isFloat16())
    return float16(getPayload<uint16_t>(2));
  if (isFloat32())
    return float16(getPayload<float>(4));
  if (isFloat64())
    return float16((float)getPayload<double>(8));
  if (isInt())
    return float16((float)getSignedInt());

  return float16();
}
float Entry::getFloat32() const {
  if (matchType(0b000100, 0b111111))
    return 0.0f;
  if (isFloat16())
    return (float)float16(getPayload<uint16_t>(2));
  if (isFloat32())
    return getPayload<float>(4);
  if (isFloat64())
    return (float)getPayload<double>(8);
  if (isInt())
    return (float)getSignedInt();

  return 0.0f;
}
double Entry::getFloat64() const {
  if (matchType(0b000100, 0b111111))
    return 0.0;
  if (isFloat16())
    return (double)float16(getPayload<uint16_t>(2));
  if (isFloat32())
    return (double)getPayload<float>(4);
  if (isFloat64())
    return getPayload<double>(8);
  if (isInt())
    return (double)getSignedInt();

  return 0.0;
}

uint8_t Entry::getBytes(uint8_t *bytes) const {
  if (matchType(0b000011)) {
    uint8_t len = getPayload<uint8_t>(1);
    std::memcpy(bytes, getPayloadBuf() + 1, len);
    return len;
  }
  if (matchType (0b001000, 0b111000)) {
    uint8_t len = getType() & 0b000111;
    std::memcpy(bytes, getPayloadBuf(), len);
    return len;
  }
  return 0;
}

uint8_t Entry::getString(char *str) const {
  uint8_t len = getBytes(reinterpret_cast<uint8_t*>(str));
  str[len] = '\0';
  return len;
}

const Packet Entry::getPacket() const {
  if (isPacket()) return Packet::decode(getPayloadBuf());
  else            return Packet::null();
}

const SubEntries Entry::getStruct() const {
  return SubEntries(entries_, ptr_ + entry_type_size, entries_.buf_, entries_.buf_size_);
}


void Entry::remove() {
  if (*this) {
    entries_.resize(ptr_, 0, entry_type_size + size());
  }
}

bool Entry::setNull() {
  if (!setSize(0))
    return false;
  setType(0b000000);
  return true;
}

bool Entry::setInt(const uint8_t *bytes, uint8_t size, bool is_negative) {
  if (size == 0) {
    if (!setSize(0)) return false;
    setType(0b100000);
  }
  else if (size == 1 && !is_negative && bytes[0] < 32) {
    if (!setSize(0)) return false;
    setType(0b100000 | bytes[0]);
  }
  else {
    if (!setSize(size)) return false;
    if (is_negative) setType(0b011000 | (size - 1));
    else             setType(0b010000 | (size - 1));
    setPayload(bytes, size);
  }
  return true;
}

bool Entry::setFloat16(float value) {
  if (value == 0.0f) {
    if (!setSize(0))
      return false;
    setType(0b000100);
  } else {
    if (!setSize(2))
      return false;
    setType(0b000101);
    float16 value16(value);
    uint16_t raw = value16.getRaw();

    setPayload(raw, 2);
  }
  return true;
}

bool Entry::setFloat32(float value) {
  if (value == 0.0f) {
    if (!setSize(0))
      return false;
    setType(0b000100);
  }
  else {
    if (!setSize(4))
      return false;
    setType(0b000110);
    setPayload(value, 4);
  }
  return true;
}

bool Entry::setFloat64(double value) {
  if (sizeof(double) != 8)
    return setFloat32(value);

  if (value == 0.0f) {
    if (!setSize(0))
      return false;
    setType(0b000100);
  } else {
    if (!setSize(8))
      return false;
    setType(0b000111);
    setPayload(value, 8);
  }
  return true;
}

bool Entry::setBytes(const uint8_t *bytes, uint8_t length) {
  if (length <= 7) {
    if (!setSize(length))
      return false;
    setType(0b001000 | length);
    setPayload(bytes, length);
  }
  else {
    if (!setSize(length + 1))
      return false;
    setType(0b000011);
    setPayload(length, 1);
    setPayload(bytes, length, 1);
  }
  return true;
}

bool Entry::setString(const char *str) {
  return setBytes(reinterpret_cast<const uint8_t*>(str), std::strlen(str));
}

SubEntries Entry::setStruct() {
  if (setSize(1)) {
    setType(0b000001);
    setPayload(1, 1);
  } 
  return SubEntries(entries_, ptr_ + entry_type_size, entries_.buf_, entries_.buf_size_);
}

bool Entry::setPacket(const Packet& packet) {
  if (!setSize(packet.size())) return false;
  setType(0b000010);
  setPayload(packet.encode(), packet.size());
  return true; 
}

bool Entry::setSize(uint8_t size_new) {
  return entries_.resize(ptr_ + entry_type_size, size_new, size());
}

const uint8_t* Entry::getPayloadBuf() const {
  return entries_.buf_ + ptr_ + entry_type_size;
}
uint8_t* Entry::getPayloadBuf() {
  return entries_.buf_ + ptr_ + entry_type_size;
}


EntriesIterator &EntriesIterator::operator++() {
  if (ptr_ + (**this).size() + entry_type_size < entries_.offset() + entries_.size())
    ptr_ += (**this).size() + entry_type_size;
  else ptr_ = entries_.offset() + entries_.size();
  return *this;
}

EntriesConstIterator &EntriesConstIterator::operator++() {
  if (ptr_ + (**this).size() + entry_type_size < entries_.offset() + entries_.size())
    ptr_ += (**this).size() + entry_type_size;
  else ptr_ = entries_.offset() + entries_.size();
  return *this;
}


EntriesIterator& EntriesIterator::find(const char name[2]) {
  while (*this != entries_.end() && (**this).name() == name) ++(*this);
  return *this;
}

EntriesConstIterator& EntriesConstIterator::find(const char name[2]) {
  while (*this != entries_.end() && (**this).name() == name) ++(*this);
  return *this;
}

Entries::iterator Entries::find(const char name[2]) {
  iterator itr = begin();
  return itr.find(name);
}

Entry EntriesIterator::insert(const char name[2]) {
  Entry e(entries_, ptr_);
  if (entries_.resize(ptr_, entry_type_size, 0)) {
    e.name() = name;
    e.setType(0b000000);
  }
  return e;
}

Entries::const_iterator Entries::find(const char name[2]) const {
  const_iterator itr = begin();
  return itr.find(name);
}


Entries::iterator Entries::at(unsigned n) {
  iterator itr = begin();
  for (int i = 0; i < n; i++) ++itr;
  return itr;
}

Entries::const_iterator Entries::at(unsigned n) const {
  const_iterator itr = begin();
  for (int i = 0; i < n; i++) ++itr;
  return itr;
}

Entry Entries::append(const char name[2]) {
  return end().insert(name);
}


void Entries::clear() {
  printf("CLEAR %d %d %d\n", offset(), header_size(), size());
  resize(offset() + header_size(), 0, size() - header_size());
}

Packet &Packet::command(uint8_t packet_id, uint8_t component_id) {
  buf_[1] = packet_id & ~(packet_type_mask);
  buf_[2] = component_id;
  buf_[3] = unit_id_local;
  resize(0, 4, buf_[0]);
  return *this;
};

Packet &Packet::command(uint8_t packet_id, uint8_t component_id,
                        uint8_t origin_unit_id, uint8_t dest_unit_id, uint16_t sequence) {
  unsigned header_size_old = header_size();
  buf_[1] = packet_id & ~(packet_type_mask);
  buf_[2] = component_id;
  buf_[3] = origin_unit_id;
  buf_[4] = dest_unit_id;
  buf_[5] = sequence & 0xFF;
  buf_[6] = sequence >> 8;
  resize(0, 7, buf_[0]);
  return *this;
};

Packet &Packet::telemetry(uint8_t packet_id, uint8_t component_id) {
  unsigned header_size_old = header_size();
  buf_[1] = packet_id | packet_type_mask;
  buf_[2] = component_id;
  buf_[3] = unit_id_local;
  buf_[0] = header_size();
  resize(0, 4, buf_[0]);
  return *this;
};

Packet &Packet::telemetry(uint8_t packet_id, uint8_t component_id,
                          uint8_t origin_unit_id, uint8_t dest_unit_id, uint16_t sequence) {
  unsigned header_size_old = header_size();
  buf_[1] = packet_id | packet_type_mask;
  buf_[2] = component_id;
  buf_[3] = origin_unit_id;
  buf_[4] = dest_unit_id;
  buf_[5] = sequence & 0xFF;
  buf_[6] = sequence >> 8;
  buf_[0] = header_size();
  resize(0, 7, buf_[0]);
  return *this;
};

Packet& Packet::operator=(const Packet& packet) {
  buf_ = packet.buf_;
  buf_size_ = packet.buf_size_;
  ref_change_ = packet.ref_change_;
  if (ref_change_ != nullptr) ref_change_(*this, +1);
  return *this;
}

Packet& Packet::operator=(Packet&& packet) {
  buf_ = packet.buf_;
  buf_size_ = packet.buf_size_;
  ref_change_ = packet.ref_change_;
  return *this;
}

bool Packet::copyPayload(const Packet& from) {
  if (buf_size_ - header_size() < from.size() - from.header_size()) return false;

  buf_[0] = header_size() + from.size() - from.header_size();
  std::memcpy(buf_ + header_size(), from.buf_ + header_size(),
              from.size() - from.header_size());
  return true;
}

bool Packet::copy(const Packet& from) {
  if (buf_size_ < from.size()) return false;

  std::memcpy(buf_, from.buf_, from.size() + 1);
  return true;
}

bool Packet::resize(uint8_t ptr, uint8_t size_from_ptr, uint8_t size_from_ptr_old) {
  // printf("RESIZE P %d %d %d %d\n", ptr, size_from_ptr, size_from_ptr_old, buf_size_);
  if (size() + size_from_ptr - size_from_ptr_old > buf_size_) return false;
  std::memcpy(buf_ + ptr + size_from_ptr, buf_ + ptr + size_from_ptr_old, size() - ptr - size_from_ptr_old);
  buf_[0] += size_from_ptr - size_from_ptr_old;
  return true;
}

bool SubEntries::resize(uint8_t ptr, uint8_t size_from_ptr, uint8_t size_from_ptr_old) {
  // printf("RESIZE S %d %d %d %d %d %d\n", offset(), size(), ptr, size_from_ptr, size_from_ptr_old, buf_size_);
  if (size() + size_from_ptr - size_from_ptr_old + offset() > buf_size_) return false;
  if (!parent_.resize(ptr, size_from_ptr, size_from_ptr_old)) return false;
  buf_[offset()] += size_from_ptr - size_from_ptr_old;
  return true;
}


} // namespace wccp
